### PR TITLE
feat (projects): auth session limit

### PR DESCRIPTION
**Pull request by lohanidamodar**
Add endpoint on project to set max number of sessions for project users.

## What does this PR do?

- introduces feature to set max session limits per project

### UI
image

## Test Plan

- New test added to `ProjectConsoleClientTest`

## Related PRs and Issues

- UI - https://github.com/appwrite/console/pull/215

### Have you added your change to the [Changelog](https://github.com/appwrite/appwrite/blob/master/CHANGES.md)?

(The CHANGES.md file tracks all the changes that make it to the `main` branch. Add your change to this file in the following format)
- One line description of your PR [#pr_number](Link to your PR)

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

YES

### CODE REVIEW

File: app/init.php [hunk 1]
```diff
@@ -84,6 +84,7 @@
 const APP_PAGING_LIMIT = 12;
 const APP_LIMIT_COUNT = 5000;
 const APP_LIMIT_USERS = 10000;
+const APP_LIMIT_USER_SESSIONS = 100;
```
Comment (eldadfux) [inline app/init.php:87]: I feel we should aim lower. Maybe 20.

File: app/controllers/shared/api.php [hunk 2]
```diff
@@ -318,6 +318,39 @@
         }
     });
 
+App::shutdown()
+    ->groups(['session'])
```
Comment (eldadfux) [inline app/controllers/shared/api.php:322]: Can we add a description to each hook explaining its responsibility?

File: app/controllers/shared/api.php [hunk 3]
```diff
@@ -318,6 +318,39 @@
...
+    ->inject('utopia')
+    ->inject('request')
+    ->inject('response')
+    ->inject('project')
+    ->inject('dbForProject')
+    ->action(function (App $utopia, Request $request, Response $response, Document $project, Database $dbForProject) {
+        $sessionLimit = $project->getAttribute('auths', [])['maxSessions'] ?? APP_LIMIT_USER_SESSIONS;
+        $session = $response->getPayload();
```
Comment (eldadfux) [inline app/controllers/shared/api.php:330]: Does this works well for OAuth sessions as well?

File: src/Appwrite/Utopia/Response/Model/Project.php [hunk 4]
```diff
@@ -114,6 +114,12 @@ public function __construct()
                 'default' => 0,
                 'example' => 100,
             ])
+            ->addRule('authSessionsLimit', [
+                'type' => self::TYPE_INTEGER,
+                'description' => 'Max sessions allowed per user. 100 maximum.',
+                'default' => 100,
```
Comment (eldadfux) [inline src/Appwrite/Utopia/Response/Model/Project.php:120]:
```suggestion
                'default' => 100,
```

Lower...

**Review by eldadfux — CHANGES_REQUESTED** [review]

File: app/init.php (hunk 1 above)
Comment (lohanidamodar) [reply app/init.php:87]: This is the maximum ever allowed? Are we sure to keep it low as 20?, default is 10 hardcoded

File: app/controllers/shared/api.php (hunk 3 above)
Comment (lohanidamodar) [reply app/controllers/shared/api.php:330]: it should as I added the group in the endpoint that actually creates the session object

File: app/init.php (hunk 1 above)
Comment (lohanidamodar) [reply app/init.php:87]: this is also used the in the range validation of the endpoint `1-100` that's the max allowed and default. Do we keep separate constant for max allowed value and default? or make it lower as 10-20 ?

File: app/controllers/shared/api.php (hunk 2 above)
Comment (lohanidamodar) [reply app/controllers/shared/api.php:322]: I'll add to this one, but others are handling so many things

Comment (eldadfux) [reply app/controllers/shared/api.php:322]: Agree. Eventually I think we should start splitting them to have one responsibility.

File: src/Appwrite/Utopia/Response/Model/Project.php (hunk 4 above)
Comment (eldadfux) [reply src/Appwrite/Utopia/Response/Model/Project.php:120]: Let's change the default to 10 as well

**Review by eldadfux — APPROVED** [review]

**Pull request merged**